### PR TITLE
Improve BanyanDB measure schema 

### LINKED ISSUE #10105: [BanyanDB] Remove ID tag from the measure’s schema.
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/skywalking/issues?q=is%3Aissue) and found no similar feature requirement.

### Description

OAP’s logical id is used to fetch arbitrary data from the storage layer. In BanyanDB’s scenario, we found a chance to remove this intermediate transformation. 

In a typical scenario, OAP combines *entity_id + timebucket* to logical id. When the data is written, BanyanDB will create an inverted index to map the id to an internal item id which is *series_id + timestamp*. *series_id* is encoded from *entity_id*, and *timestamp* is encoded from *timebucket*. So we can remove the logical id to use the original combination to fetch data.

Based on the above, we could remove the ID tag from the measure’s schema in the metric creation phase, then transform *by id* queries to *by entity_id +timebucket*.

### Use case

_No response_

### Related issues

_No response_

### Are you willing to submit a PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by hanahmily**
## Remove OAL and MAL generated measures' ID tag

The previous default `ID` tag is removed from measures created by OAL and MAL. `multiGet` and `metricsQuery` methods will leverage *series id* and *time bucket* to retrieve measure data.

## Introduce `Banyandb.MeasureField` annotation

The new annotation marks a column as a measure field instead of a tag. Now a typical measure looks like below:

```yaml
entity:
    tagNames:
    - entity_id
  fields:
  - compressionMethod: COMPRESSION_METHOD_ZSTD
    encodingMethod: ENCODING_METHOD_UNSPECIFIED
    fieldType: FIELD_TYPE_STRING
    name: value
  - compressionMethod: COMPRESSION_METHOD_ZSTD
    encodingMethod: ENCODING_METHOD_GORILLA
    fieldType: FIELD_TYPE_INT
    name: precision
  - compressionMethod: COMPRESSION_METHOD_ZSTD
    encodingMethod: ENCODING_METHOD_UNSPECIFIED
    fieldType: FIELD_TYPE_STRING
    name: dataset
  interval: 1m
  metadata:
    group: measure-minute
    name: service_percentile_minute
  tagFamilies:
  - name: default
    tags:
    - indexedOnly: false
      name: entity_id
      type: TAG_TYPE_STRING
```

The measure contains a single tag named `entity_id`, which belongs to the `entity`.  No index is created. 

Furthermore, if we implement #10229, the `dataset` and `value` fields could get encoded.

## Indexed measures remained

```
 name: cache_access_sla_minute-index-rule-binding
    name: cache_access_sla_minute
    name: cache_read_sla_minute-index-rule-binding
    name: cache_read_sla_minute
    name: cache_write_sla_minute-index-rule-binding
    name: cache_write_sla_minute
    name: database_access_sla_minute-index-rule-binding
    name: database_access_sla_minute
    name: endpoint_relation_server_side_minute-index-rule-binding
    name: endpoint_relation_server_side_minute
    name: endpoint_relation_sla_minute-index-rule-binding
    name: endpoint_relation_sla_minute
    name: endpoint_sla_minute-index-rule-binding
    name: endpoint_sla_minute
    name: mq_endpoint_consume_sla_minute-index-rule-binding
    name: mq_endpoint_consume_sla_minute
    name: mq_endpoint_produce_sla_minute-index-rule-binding
    name: mq_endpoint_produce_sla_minute
    name: mq_service_consume_sla_minute-index-rule-binding
    name: mq_service_consume_sla_minute
    name: mq_service_produce_sla_minute-index-rule-binding
    name: mq_service_produce_sla_minute
    name: process_relation_client_side_minute-index-rule-binding
    name: process_relation_client_side_minute
    name: process_relation_server_side_minute-index-rule-binding
    name: process_relation_server_side_minute
    name: service_instance_relation_client_call_sla_minute-index-rule-binding
    name: service_instance_relation_client_call_sla_minute
    name: service_instance_relation_client_side_minute-index-rule-binding
    name: service_instance_relation_client_side_minute
    name: service_instance_relation_server_call_sla_minute-index-rule-binding
    name: service_instance_relation_server_call_sla_minute
    name: service_instance_relation_server_side_minute-index-rule-binding
    name: service_instance_relation_server_side_minute
    name: service_instance_sla_minute-index-rule-binding
    name: service_instance_sla_minute
    name: service_relation_client_call_sla_minute-index-rule-binding
    name: service_relation_client_call_sla_minute
    name: service_relation_client_side_minute-index-rule-binding
    name: service_relation_client_side_minute
    name: service_relation_server_call_sla_minute-index-rule-binding
    name: service_relation_server_call_sla_minute
    name: service_relation_server_side_minute-index-rule-binding
    name: service_relation_server_side_minute
    name: service_sla_minute-index-rule-binding
    name: service_sla_minute
```

`relation` metrics have `source` and `dest` tags that need to be indexed. 

`sla` metrics are from `PercentMetrics` whose `mark` column doesn't marked as `storageOnly` https://github.com/apache/skywalking/blob/f3b567160c/oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/analysis/metrics/PercentMetrics.java#L46

@wu-sheng ^^ is this intended or by mistake?

## Add `BanyanDB.StoreIDTag` annotation

Even though we remove most `ID` tags, the process traffic still needs the `ID` tag. There is a scenario to query a process by ID. But the process id is a hashed result and can not be analyzed to a structured object. So I add this new annotation to store the id as a tag for querying.

@mrproliu @wu-sheng  The banyandb also encodes `name` and `instanceId` to internal ID(SeriesID) by the xxHash algorithm. We could pass the processId to the server as the internal id. The server will skip the encoding process by storing the ingested id directly. Do you think it makes sense?

- [x] If this pull request closes/resolves/fixes an existing issue, replace the issue number. Closes #10105 .
- [x] Update the [`CHANGES` log](https://github.com/apache/skywalking/blob/master/docs/en/changes/changes.md).

### CODE REVIEW

**Comment by wu-sheng** [comment]
> sla metrics are from PercentMetrics whose mark column doesn't marked as storageOnly https://github.com/apache/skywalking/blob/f3b567160c/oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/analysis/metrics/PercentMetrics.java#L46

This is a bug, please fix as well.

File: oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/source/ScopeDefaultColumn.java [hunk 1]
```diff
@@ -34,14 +34,16 @@ public class ScopeDefaultColumn {
     private String columnName;
     private Class<?> type;
     private boolean isID;
+    private boolean banyandbNoIndexing;
     private int length;
 
-    public ScopeDefaultColumn(String fieldName, String columnName, Class<?> type, boolean isID, int length) {
+    public ScopeDefaultColumn(String fieldName, String columnName, Class<?> type, boolean isID, int length, boolean banyandbNoIndexing) {
```
Comment (wu-sheng) [inline oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/source/ScopeDefaultColumn.java:40]: This seems an unreasonable change. `ScopeDefaultColumn` is never storage related.

Comment (hanahmily) [reply oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/source/ScopeDefaultColumn.java:40]: `service_id` is defined by `ScopeDefaultColumn`.  IIUC, `ScopeDefaultColumn.columnName` defines the column name in the storage layer. Essentially, I tend to mark `service_id` as a non-indexed tag. Adding a new field seems relevant.

File: oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/storage/model/StorageModels.java [hunk 2]
```diff
@@ -190,7 +194,7 @@ private void retrieval(final Class<?> clazz,
                 ElasticSearchExtension elasticSearchExtension = new ElasticSearchExtension(
                     elasticSearchAnalyzer == null ? null : elasticSearchAnalyzer.analyzer(),
                     elasticSearchColumn == null ? null : elasticSearchColumn.columnAlias(),
-                    keywordColumn == null ? false : true
+                        keywordColumn != null
```
[the remark this answers is not in the document]
Comment (hanahmily) [reply oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/storage/model/StorageModels.java:197]: `checkStyle` reports this line should be simplified as this.

File: oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/source/ScopeDefaultColumn.java (hunk 1 above)
Comment (hanahmily) [reply oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/source/ScopeDefaultColumn.java:40]: `isID` is storage-related, right?

Comment (hanahmily) [reply oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/source/ScopeDefaultColumn.java]: Dropped it.

File: oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/source/ScopeDefaultColumn.java [hunk 3]
```diff
@@ -52,14 +54,22 @@ public ScopeDefaultColumn(String fieldName, String columnName, Class<?> type, bo
...
          */
         int length() default 256;
+
+
+        /**
+         * Active indexing for a column(tag) in BanyanDB.
+         *
+         * @return false: Indexing the column, true: No indexing the column.
+         */
+        boolean requiredNoIndexingForBanyanDB() default false;
```
[the remark this answers is not in the document]
Comment (hanahmily) [reply oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/source/ScopeDefaultColumn.java]: Removed

File: oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/storage/model/StorageModels.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (wu-sheng) [reply oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/storage/model/StorageModels.java:197]: Maybe that is some of your local change. I tried formating locally, nothing changed.

File: oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/storage/annotation/BanyanDB.java [hunk 4]
```diff
@@ -131,4 +131,22 @@ enum IndexType {
...
+    @interface MeasureField {
+    }
+
+    /**
+     * StoreIDTag indicates a metric store its ID as a tag for searching.
+     * @Since 9.4.0
+     */
+    @Target({ElementType.TYPE})
+    @Retention(RetentionPolicy.RUNTIME)
+    @interface StoreIDTag {
```
Comment (wu-sheng) [inline oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/storage/annotation/BanyanDB.java:150]:
```suggestion
    @interface StoreIDAsTag {
```

I think this name is more accurate.

File: oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/storage/annotation/BanyanDB.java [hunk 5]
```diff
@@ -131,4 +131,22 @@ enum IndexType {
     @interface TimestampColumn {
         String value();
     }
+
+    /**
+     * MeasureField defines a column as a measure's field.
```
Comment (wu-sheng) [inline oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/storage/annotation/BanyanDB.java:136]: I think we need to explain if it is not annotated, what a column would be processed.
Such as
1. storageOnly=true without annotated
2. indexOnly=true without annotated
3. storageOnly=true with annotated
4. indexOnly=true with annotated
5. storageOnly=false and indexOnly=false with annotated
5. indexOnly=false and indexOnly=false without annotated

Comment (hanahmily) [reply oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/storage/annotation/BanyanDB.java:136]: I will update the comments.

File: oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/storage/model/StorageModels.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (hanahmily) [reply oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/storage/model/StorageModels.java:197]: My IDE might suggest this change. But simple is better than complex. This tweak makes sense to me.

[the remark this answers is not in the document]
Comment (hanahmily) [reply oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/storage/model/StorageModels.java]: Done

File: oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/storage/annotation/BanyanDB.java (hunk 5 above)
Comment (hanahmily) [reply oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/storage/annotation/BanyanDB.java:136]: Add more details to comments.

File: oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/storage/model/BanyanDBModelExtension.java [hunk 6]
```diff
@@ -37,4 +37,12 @@ public class BanyanDBModelExtension {
     @Setter
     private String timestampColumn;
 
+    /**
+     * shouldStoreIDTag indicates whether a metric store its ID as a tag.
```
Comment (wu-sheng) [inline oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/storage/model/BanyanDBModelExtension.java:41]:
```suggestion
     * shouldStoreIDTag indicates whether a metric stores its ID as a tag.
```

File: oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/storage/annotation/BanyanDB.java [hunk 7]
```diff
@@ -131,4 +131,29 @@ enum IndexType {
...
+     *
+     * Annotated: the column is a measure field.
+     * Unannotated: the column is a measure tag.
+     *   storageOnly=true: the column is a measure tag which is not indexed.
+     *   storageOnly=false: the column is a measure tag which is indexed.
+     *     indexOnly=true: the column is a measure tag which is indexed, but not stored.
+     *     indexOnly=false: the column is a measure tag which is indexed and stored.
```
Comment (wu-sheng) [inline oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/storage/annotation/BanyanDB.java:143]:
```suggestion
     *   storageOnly=true: the column is a measure tag that is not indexed.
     *   storageOnly=false: the column is a measure tag that is indexed.
     *   indexOnly=true: the column is a measure tag that is indexed, but not stored.
     *   indexOnly=false: the column is a measure tag that is indexed and stored.
```

File: oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/source/ScopeDefaultColumn.java [hunk 8]
```diff
@@ -52,7 +52,7 @@ public ScopeDefaultColumn(String fieldName, String columnName, Class<?> type, bo
         /**
          * Dynamic active means this column is only activated through core setting explicitly.
          *
-         * @return
+         * @return false: Inactive dynamic column. true: Active dynamic column.
```
Comment (wu-sheng) [inline oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/source/ScopeDefaultColumn.java:55]: This comment seems a little strange. 

```suggestion
         * @return FALSE: this column is not going to be added to the final generated metric as a column.
         *         TRUE: this column could be added as a column if core/activeExtraModelColumns == true.
```

**Review by wu-sheng — COMMENTED** [review]
Others are good to me. There are some nits mentioned above.

**Comment by wu-sheng** [comment]
@wankai123 There are some SeriesID changes, could you check?

**Comment by wu-sheng** [comment]
@hanahmily I think this resolves #10033 too.

**Review by wankai123 — APPROVED** [review]
LGTM

File: oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/analysis/manual/service/ServiceTraffic.java [hunk 9]
```diff
@@ -92,7 +92,7 @@ public class ServiceTraffic extends Metrics {
     @Setter
     @Getter
     @Column(columnName = LAYER)
-    @BanyanDB.SeriesID(index = 1)
+    @BanyanDB.SeriesID(index = 0)
```
Comment (wu-sheng) [inline oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/analysis/manual/service/ServiceTraffic.java:95]: What is the reason of the layer first?

Comment (hanahmily) [reply oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/analysis/manual/service/ServiceTraffic.java:95]: Searching SeriesID supports prefix scanning. OAP uses finding by `layer` a lot. Based on that, moving the `layer` to the first place will improve the search performance a bit.

Comment (wu-sheng) [reply oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/analysis/manual/service/ServiceTraffic.java:95]: Make sense. Let's add a little comments to seriesID index.
We should indicate the index matters for prefix searching.

Comment (hanahmily) [reply oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/analysis/manual/service/ServiceTraffic.java:95]: Done

**Review by wu-sheng — APPROVED** [review]
LGTM

**Comment by kezhenxu94** [comment]
This PR reverts the UI module unexpectedly, https://github.com/apache/skywalking/pull/10249

**Pull request merged**